### ISSUE TITLE
libelf dependency error while compiling mesa

### ISSUE THREAD

**Issue by Decatf** [postmarketOS/pmbootstrap#150]
I'm trying to compile mesa from source. I've taken [APKBUILD and related files](https://git.alpinelinux.org/cgit/aports/tree/main/mesa) from upstream Alpine aports and put them into the aports folder. Is this the way to do it?

```
(016088) [14:05:36] *** (1/5) PREPARE NATIVE CHROOT ***
(016088) [14:05:36] Calculate depends of packages ['cryptsetup', 'util-linux', 'e2fsprogs', 'parted'], arch: x86_64
(016088) [14:05:36] ERROR: Could not resolve variable $depends_dev in APKBUILD of mesa
(016088) [14:05:36] Run 'pmbootstrap log' for details.
(016088) [14:05:36] See also: <https://postmarketos.org/troubleshooting>
(016088) [14:05:36] Traceback (most recent call last):
  File "./pmbootstrap.py", line 81, in main
    pmb.install.install(args)
  File "/home/r/pmbootstrap/pmb/install/install.py", line 85, in install
    build=False)
  File "/home/r/pmbootstrap/pmb/chroot/apk.py", line 183, in install
    strict=True)
  File "/home/r/pmbootstrap/pmb/parse/depends.py", line 87, in recurse
    aport = pmb.build.find_aport(args, pkgname, False)
  File "/home/r/pmbootstrap/pmb/build/other.py", line 48, in find_aport
    apkbuild = pmb.parse.apkbuild(args, path_current)
  File "/home/r/pmbootstrap/pmb/parse/apkbuild.py", line 128, in apkbuild
    ret = replace_variables(ret)
  File "/home/r/pmbootstrap/pmb/parse/apkbuild.py", line 50, in replace_variables
    apkbuild["pkgname"])
RuntimeError: Could not resolve variable $depends_dev in APKBUILD of mesa
```

The current Alpine Linux mesa binary is compiled with `-mpu=neon`. It appears this was done because (in a recent change) mesa does runtime CPU feature detection instead of compile time flags. Xwayland is failing to due undefined opcodes on my non-NEON device. I want to get mesa built from source in order to debug this.

**Comment by ollieparanoid** [comment]
That is exactly the way to do it. The `APKBUILD` parsing (`pmb.parse.apkbuild`) is very rudimentary on purpose (otherwise we'd have to re-implement a shell parser), so you might need to add some variables there/in `pmb/config/__init__.py`, such as the `depends_dev` variable. If you need any help with that, let me know :)

Great finding with Xwayland - this might just be the reason why it fails for the `lg-mako` (and other devices?)!

(I'm hanging out in the chat currently, there I could help you more directly.)

**Comment by Decatf** [comment]
It's unlikely this problem is related to other devices. The mako CPU has NEON. The Nvidia Tegra 2 devices are probably the most popular Android device with a CPU that lacks NEON. There are a handful of Android Honeycomb/ICS tablets that were made with this chip.

This CPU feature is optional but it's pretty much de facto standard for every ARM based Android device. Any device that also does not have NEON is going to be some obscure device or a very very old one (pre year 2011-ish).

**Comment by ollieparanoid** [comment]
Thanks for the explanation :+1:

**Comment by Decatf** [comment]
After adding `depends_dev` variable. The build process gets to here:

```
(029014) [20:31:15] *** (1/5) PREPARE NATIVE CHROOT ***
(029014) [20:31:15] Calculate depends of packages ['cryptsetup', 'util-linux', 'e2fsprogs', 'parted'], arch: x86_64
(029014) [20:31:15] *** (2/5) CREATE DEVICE ROOTFS (samsung-p4wifi) ***
(029014) [20:31:15] (rootfs_samsung-p4wifi) % apk update
fetch https://nl.alpinelinux.org/alpine/edge/main/armhf/APKINDEX.tar.gz
fetch https://nl.alpinelinux.org/alpine/edge/community/armhf/APKINDEX.tar.gz
fetch https://nl.alpinelinux.org/alpine/edge/testing/armhf/APKINDEX.tar.gz
user  [/home/user/packages/user]
v3.6.0-1850-g69bf70aaa3 [https://nl.alpinelinux.org/alpine/edge/main]
v3.6.0-1849-ga52a40744b [https://nl.alpinelinux.org/alpine/edge/community]
v3.6.0-1842-g8409258a64 [https://nl.alpinelinux.org/alpine/edge/testing]
OK: 10692 distinct packages available
(029014) [20:31:19] Calculate depends of packages dict_keys(['so:libintl.so.8', 'mcookie', 'graphite2', 'so:libsudo_util.so.0', 'so:libgobject-2.0.so.0', 'so:libp11-kit.so.0', 'mcpp', 'apk-tools', 'so:libhogweed.so.4', 'so:libgmp.so.10', 'so:libcupsimage.so.2', 'libnl3', 'musl', 'findmnt', 'so:libunwind-ptrace.so.0', 'so:liburcu-bp.so.6', 'so:libevent_pthreads-2.1.so.6', 'so:libcryptsetup.so.4', 'mosh', 'so:libnl-genl-3.so.200', 'perl', 'busybox', 'encodings', 'gmp', 'so:system_group.so', 'so:libpcreposix.so.0', 'alpine-conf', 'so:libffi.so.6', 'so:liburcu-cds.so.6', 'libevent', 'eudev-libs', 'cryptsetup', 'so:libxcb-xkb.so.1', 'so:libGL.so.1', 'so:libmount.so.1', 'so:libXcursor.so.1', 'pcsc-lite-libs', 'so:libcrypto.so.41', 'so:libxcb-glx.so.0', 'so:libpcsclite.so.1', 'mcpp-libs', 'so:libjson-c.so.2', 'libgcc', 'so:libxcb.so.1', 'so:libwayland-server.so.0', 'avahi-libs', 'so:libdevmapper.so.1.02', 'xorg-server-xwayland', 'udev-init-scripts', 'so:libunwind.so.8', 'hicolor-icon-theme', 'libelf', 'so:libxcb-composite.so.0', 'so:libunwind-coredump.so.0', 'so:libturbojpeg.so.0', 'openssh', 'mesa-glapi', 'so:libwayland-egl.so.1', 'so:libpanelw.so.6', 'libwebp', 'libweston', 'so:libva-tpi.so.1', 'so:libcupscgi.so.1', 'so:libXcomposite.so.1', 'htop', 'libstdc++', 'so:libxcb-res.so.0', 'so:libpixman-1.so.0', 'dbus-libs', 'libxfixes', 'so:libxcb-dri2.so.0', 'so:libunwind-setjmp.so.0', 'libbsd', 'so:libtasn1.so.6', 'atk', 'ncurses-libs', 'libxau', 'weston-backend-x11', 'busybox-suid', 'mkfontscale', 'so:libva-x11.so.1', 'so:libweston-desktop-2.so.0', 'so:liburcu-common.so.6', 'so:libgbm.so.1', 'mosh-client', 'so:libreadline.so.7', 'mesa-gl', 'so:libblkid.so.1', 'so:libglapi.so.0', 'so:libXtst.so.6', 'libxinerama', 'so:libpangoft2-1.0.so.0', 'so:libEGL.so.1', 'so:libatk-bridge-2.0.so.0', 'so:libpangoxft-1.0.so.0', 'so:libgdbm.so.4', 'json-c', 'so:libXt.so.6', 'libsmartcols', 'so:libXdmcp.so.6', 'so:libunistring.so.2', 'so:libinput.so.10', 'so:libsqlite3.so.0', 'cairo-gobject', 'libressl2.5-libcrypto', 'so:libLLVM-4.0.so', 'llvm4-libs', 'so:libprotoc.so.11', 'so:libXext.so.6', 'mesa-gles', 'so:libgdbm_compat.so.4', 'so:libtiff.so.5', 'gnutls', 'libxtst', 'libxext', 'so:libdrm_radeon.so.1', 'xmodmap', 'ncurses-widec-libs', 'so:libxcb-shm.so.0', 'so:libunwind-arm.so.8', 'so:libpcscspy.so.0', 'libxdamage', 'so:liburcu-mb.so.6', 'protobuf', 'so:libdrm_freedreno.so.1', 'so:libformw.so.6', 'gtk-update-icon-cache', 'so:libmpathpersist.so.0', 'so:libevent_openssl-2.1.so.6', 'so:liburcu-qsbr.so.6', 'mesa-libwayland-egl', 'so:libprotobuf.so.11', 'so:libXi.so.6', 'so:libcairo-gobject.so.2', 'font-cursor-misc', 'at-spi2-core', 'font-alias', 'so:libxcb-xf86dri.so.0', 'so:libdrm.so.2', 'so:libX11.so.6', 'nettle', 'pcre', 'so:libfontconfig.so.1', 'libaio', 'so:libxcb-present.so.0', 'so:libelf.so.0', 'wpa_supplicant', 'so:libdbus-1.so.3', 'so:libssl.so.43', 'cups-libs', 'libintl', 'so:libbsd.so.0', 'openrc', 'so:libbz2.so.1', 'so:libavahi-client.so.3', 'so:liburcu-signal.so.6', 'so:libcupsmime.so.1', 'so:libXmuu.so.1', 'python3', 'so:libgcc_s.so.1', 'so:libfreetype.so.6', 'weston-backend-drm', 'libxdmcp', 'lddtree', 'so:libsmartcols.so.1', 'so:libva.so.1', 'libinput', 'so:libkms.so.1', 'so:libevent_core-2.1.so.6', 'device-samsung-p4wifi', 'so:libxcb-render.so.0', 'xauth', 'so:libz.so.1', 'xrdb', 'mesa-dri-swrast', 'ttf-droid', 'so:libgraphite2.so.3', 'at-spi2-atk', 'so:libgdk-3.so.0', 'so:libxcb-randr.so.0', 'libxcb', 'p11-kit', 'so:libva-drm.so.1', 'so:libcairo-script-interpreter.so.2', 'so:libmtdev.so.1', 'busybox-initscripts', 'so:libgmodule-2.0.so.0', 'so:libxcb-screensaver.so.0', 'so:libjpeg.so.8', 'mkinitfs', 'sqlite-libs', 'weston-terminal', 'so:libgthread-2.0.so.0', 'libunistring', 'so:libxcb-shape.so.0', 'ncurses-terminfo', 'so:libxcb-xinput.so.0', 'openssh-server', 'so:libprotobuf-lite.so.11', 'libc-utils', 'so:libva-glx.so.1', 'libxmu', 'so:libgtk-3.so.0', 'so:libpcre.so.1', 'libbz2', 'so:libxcb-sync.so.1', 'weston-xwayland', 'so:libpng16.so.16', 'alpine-keys', 'so:libXfont2.so.2', 'mosh-server', 'gdk-pixbuf', 'wayland', 'so:libXft.so.2', 'perl-io-tty', 'libxxf86vm', 'so:libharfbuzz.so.0', 'device-mapper-libs', 'so:group_file.so', 'so:libpython3.so', 'so:libnl-idiag-3.so.200', 'openssh-server-common', 'pango', 'so:libxcb-record.so.0', 'so:librc.so.1', 'iw', 'xinit', 'linux-samsung-p4wifi', 'libdrm', 'so:libmcpp.so.0', 'libmagic', 'weston-clients', 'so:libxcb-xfixes.so.0', 'libxkbfile', 'libuuid', 'so:libdmmp.so.0.1.0', 'libblkid', 'so:libXmu.so.6', 'libtasn1', 'so:libnl-nf-3.so.200', 'weston', 'libmount', 'libjpeg-turbo', 'pixman', 'so:libnettle.so.6', 'so:libncursesw.so.6', 'mkbootimg', 'so:libnl-route-3.so.200', 'readline', 'fontconfig', 'so:libwayland-cursor.so.0', 'so:sudoers.so', 'so:libavahi-common.so.3', 'so:libxcb-xtest.so.0', 'libxft', 'popt', 'so:libICE.so.6', 'dbus', 'so:libmultipath.so.0', 'so:libkmod.so.2', 'so:libfontenc.so.1', 'libxrandr', 'libxrender', 'so:liburcu.so.6', 'so:libmagic.so.1', 'so:libXau.so.6', 'xkeyboard-config', 'libfontenc', 'harfbuzz', 'libxshmfence', 'so:libwayland-client.so.0', 'libxml2', 'so:libxkbcommon.so.0', 'postmarketos-mkinitfs', 'so:libperl.so', 'libpng', 'multipath-tools', 'so:libGLESv1_CM.so.1', 'expat', 'mesa-egl', 'so:libeinfo.so.1', 'so:libmenuw.so.6', 'mesa-gbm', 'libxkbcommon', 'so:libXxf86vm.so.1', 'mkfontdir', 'libressl2.5-libssl', 'so:libxcb-dri3.so.0', 'userspace-rcu', 'so:libc.musl-armhf.so.1', 'libxt', 'glib', 'so:libevdev.so.2', 'busybox-extras', 'haveged', 'so:libXrender.so.1', 'cairo', 'so:libpython3.6m.so.1.0', 'so:libxcb-dpms.so.0', 'libepoxy', 'libxcomposite', 'sudo', 'so:libXdamage.so.1', 'so:libpopt.so.0', 'so:libwebpmux.so.3', 'eudev', 'tiff', 'shared-mime-info', 'so:libudev.so.1', 'so:libnl-3.so.200', 'gtk+3.0', 'so:libevent-2.1.so.6', 'kmod', 'so:libexpat.so.1', 'so:libstdc++.so.6', 'so:libpango-1.0.so.0', 'openssh-sftp-server', 'libxfont2', 'so:libxcb-xinerama.so.0', 'so:libgdk_pixbuf_xlib-2.0.so.0', 'so:libevent_extra-2.1.so.6', 'so:liblzma.so.5', 'libxcursor', 'so:libglib-2.0.so.0', 'so:libatspi.so.0', 'xz-libs', 'so:libXfixes.so.3', 'so:libdrm_nouveau.so.2', 'libsm', 'so:libcairo.so.2', 'ncurses-terminfo-base', 'weston-backend-fbdev', 'nano', 'so:libgailutil-3.so.0', 'tmux', 'so:libXrandr.so.2', 'freetype', 'xkbcomp', 'so:libxkbcommon-x11.so.0', 'libevdev', 'libice', 'libxi', 'so:libcupsppdc.so.1', 'musl-utils', 'udev', 'so:libpangocairo-1.0.so.0', 'weston-shell-desktop', 'zlib', 'so:libwebp.so.7', 'libunwind', 'font-misc-misc', 'so:libwebpdecoder.so.3', 'libx11', 'so:libgio-2.0.so.0', 'libffi', 'so:libxshmfence.so.1', 'openssh-client', 'so:libwebpdemux.so.2', 'so:libXinerama.so.1', 'libva', 'so:libmpathcmd.so.0', 'alpine-baselayout', 'so:libgnutls.so.30', 'so:libGLESv2.so.2', 'so:libxml2.so.2', 'so:libxcb-xvmc.so.0', 'so:libxcb-xv.so.0', 'so:libX11-xcb.so.1', 'scanelf', 'cryptsetup-libs', 'libweston-desktop', 'alpine-base', 'gdbm', 'mtdev', 'llvm-libs', 'so:libdrm_amdgpu.so.1', 'so:libSM.so.6', 'openssh-keygen', 'so:libatk-1.0.so.0', 'postmarketos-demos', 'so:libgdk_pixbuf-2.0.so.0', 'so:libva-egl.so.1', 'so:libxkbfile.so.1', 'so:libepoxy.so.0', 'so:libweston-2.so.0', 'so:libnl-xfrm-3.so.200', 'so:libxcb-damage.so.0', 'so:libhavege.so.1', 'postmarketos-base', 'so:libuuid.so.1', 'so:libaio.so.1', 'so:libsudo_noexec.so', 'so:libcups.so.2']), arch: armhf
(029014) [20:31:19] % cd /home/r/pmbootstrap/aports && git rev-parse --show-toplevel
/home/r/pmbootstrap
(029014) [20:31:19] % cd /home/r/pmbootstrap && git diff --name-only origin
aports/device-samsung-p4wifi/90-android.rules
aports/device-samsung-p4wifi/APKBUILD
aports/device-samsung-p4wifi/deviceinfo
aports/device-samsung-p4wifi/initfs-hook.sh
aports/linux-samsung-p4wifi/APKBUILD
aports/linux-samsung-p4wifi/config-p4wifi.armhf
aports/linux-samsung-p4wifi/config-samsung-p4wifi.armhf
aports/mesa/APKBUILD
aports/mesa/glx_ro_text_segm.patch
aports/mesa/musl-fix-includes.patch
aports/mesa/musl-fixes.patch
pmb/config/__init__.py
(029014) [20:31:19] % cd /home/r/pmbootstrap && git ls-files --others --exclude-standard
aports/linux-samsung-p4wifi/kernel.release
aports/linux-samsung-p4wifi/zImage-dtb
(029014) [20:31:19] Build is neccessary for package 'mesa': Binary package and aport have the same pkgver and pkgrel, but there are aport source files out of sync with the upstream git repository *and* these source files have a more recent 'last modified' timestamp than the binary package's build timestamp.
(029014) [20:31:19] Calculate depends of packages ['libdrm-dev', 'dri2proto', 'libx11-dev', 'libxext-dev', 'libxxf86vm-dev', 'libxdamage-dev', 'libxfixes-dev', 'libxcb-dev', 'glproto', 'dri3proto', 'presentproto', 'libxshmfence-dev', 'expat-dev', 'xextproto', 'python2', 'libxt-dev', 'makedepend', 'talloc-dev', 'py-libxml2', 'flex', 'bison', 'llvm4-dev', 'eudev-dev', 'libvdpau-dev', 'libxvmc-dev', 'gettext', 'zlib-dev', 'wayland-dev', 'libelf-dev', 'py-mako', 'autoconf', 'automake', 'libtool'], arch: armhf
(029014) [20:31:19] (buildroot_armhf) install
(029014) [20:31:19] (buildroot_armhf) % apk --no-progress add -u so:libstdc++.so.6 so:libgomp.so.1
1 errors; 446 MiB in 328 packages
(029014) [20:31:21] ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(029014) [20:31:21] NOTE: The failed command's output is above the ^^^ line in the logfile: /home/r/.local/var/pmbootstrap/log.txt
(029014) [20:31:21] ERROR: Command failed: (buildroot_armhf) % apk --no-progress add -u so:libstdc++.so.6 so:libgomp.so.1
(029014) [20:31:21] Run 'pmbootstrap log' for details.
(029014) [20:31:21] See also: <https://postmarketos.org/troubleshooting>
(029014) [20:31:21] Traceback (most recent call last):
  File "/home/r/pmbootstrap/pmb/helpers/run.py", line 45, in core
    stderr=args.logfd)
  File "/usr/lib/python3.5/subprocess.py", line 581, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['sudo', '/bin/sh', '-c', "unset $(env | cut -d= -f1); CHARSET=UTF-8 PATH=/usr/lib/ccache/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin SHELL=/bin/ash HISTFILE=~/.ash_history /usr/sbin/chroot /home/r/.local/var/pmbootstrap/chroot_buildroot_armhf sh -c 'cd /;apk --no-progress add -u so:libstdc++.so.6 so:libgomp.so.1'"]' returned non-zero exit status 1

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "./pmbootstrap.py", line 81, in main
    pmb.install.install(args)
  File "/home/r/pmbootstrap/pmb/install/install.py", line 93, in install
    pmb.chroot.apk.upgrade(args, suffix)
  File "/home/r/pmbootstrap/pmb/chroot/apk.py", line 227, in upgrade
    install(args, packages, suffix)
  File "/home/r/pmbootstrap/pmb/chroot/apk.py", line 190, in install
    args, build, arch, package, packages_installed):
  File "/home/r/pmbootstrap/pmb/chroot/apk.py", line 119, in install_is_necessary
    pmb.build.package(args, package, arch)
  File "/home/r/pmbootstrap/pmb/build/package.py", line 64, in package
    pmb.chroot.apk.install(args, apkbuild["makedepends"], suffix)
  File "/home/r/pmbootstrap/pmb/chroot/apk.py", line 212, in install
    suffix)
  File "/home/r/pmbootstrap/pmb/chroot/root.py", line 86, in root
    return_stdout, check)
  File "/home/r/pmbootstrap/pmb/helpers/run.py", line 58, in core
    raise RuntimeError("Command failed: " + log_message) from exc
RuntimeError: Command failed: (buildroot_armhf) % apk --no-progress add -u so:libstdc++.so.6 so:libgomp.so.1
```

**Comment by ollieparanoid** [comment]
`apk` doesn't really output an error message, so this looks like `apk` has already failed earlier in a `post-install` step of installing another package. I can't tell which package that was from this log, but pmbootstrap should have also failed there.

Anyway, it should continue if you run this once: `pmbootstrap chroot --suffix=buildroot_armhf -- apk fix`

(If it does not help it, please give me the log of this command again.)

**Comment by Decatf** [comment]
This the output from the shell which the above command was run.

```
r@r:~/pmbootstrap$ ./pmbootstrap.py chroot --suffix=buildroot_armhf -- apk fix
(1/1) Reinstalling libelf-dev (0.8.13-r2)
ERROR: libelf-dev-0.8.13-r2: trying to overwrite usr/include/libelf.h owned by elfutils-dev-0.168-r1.
ERROR: libelf-dev-0.8.13-r2: trying to overwrite usr/include/nlist.h owned by elfutils-dev-0.168-r1.
ERROR: libelf-dev-0.8.13-r2: trying to overwrite usr/include/gelf.h owned by elfutils-dev-0.168-r1.
ERROR: libelf-dev-0.8.13-r2: trying to overwrite usr/lib/libelf.so owned by elfutils-dev-0.168-r1.
ERROR: libelf-dev-0.8.13-r2: trying to overwrite usr/lib/libelf.a owned by elfutils-dev-0.168-r1.
1 errors; 446 MiB in 328 packages
[20:39:44] ERROR: Command failed: (buildroot_armhf) % apk fix
[20:39:44] Run 'pmbootstrap log' for details.
[20:39:44] See also: <https://postmarketos.org/troubleshooting>
r@r:~/pmbootstrap$ ./pmbootstrap.py chroot --suffix=buildroot_armhf -- apk fix
(1/1) Reinstalling libelf-dev (0.8.13-r2)
ERROR: libelf-dev-0.8.13-r2: trying to overwrite usr/include/libelf.h owned by elfutils-dev-0.168-r1.
ERROR: libelf-dev-0.8.13-r2: trying to overwrite usr/include/nlist.h owned by elfutils-dev-0.168-r1.
ERROR: libelf-dev-0.8.13-r2: trying to overwrite usr/include/gelf.h owned by elfutils-dev-0.168-r1.
ERROR: libelf-dev-0.8.13-r2: trying to overwrite usr/lib/libelf.so owned by elfutils-dev-0.168-r1.
ERROR: libelf-dev-0.8.13-r2: trying to overwrite usr/lib/libelf.a owned by elfutils-dev-0.168-r1.
1 errors; 446 MiB in 328 packages
[20:41:23] ERROR: Command failed: (buildroot_armhf) % apk fix
[20:41:23] Run 'pmbootstrap log' for details.
[20:41:23] See also: <https://postmarketos.org/troubleshooting>
```

**Comment by ollieparanoid** [comment]
Hmmm, that looks like an error in dependency resolving. Apparently both `libelf-dev` and `elfutils-dev` contain the same header file, and als `.so` file.
I'll need to look into this more, before I can fix it, so it will take some time.

As a dirty workaround hack, you could do the following:
* zap your `buildroot_armhf` chroot: `pmbootstrap zap`
* Add code [here](https://github.com/postmarketOS/pmbootstrap/blob/ec57bd7df1fb4c96118959ae2da84c6588dcc29b/pmb/chroot/apk.py#L211) to remove `libelf-dev` from the `packages_todo` list (if it does not work, try to remove `elfutils-dev`). It should be something like this (untested):

```python
packages_todo_new = []
for package in packages_todo:
    if not "libelf-dev" in package:
        packaged_todo_new.append(package)
packages_todo = packages_todo_new
```

**Comment by Decatf** [comment]
I didn't zap but I removed both of those two packages and it is compiling mesa now. The libelf-dev packaged was installed due to mesa dependency. I'm not sure what the origin of elfutils-dev dependency is though. If it's not from some other default package then it probably came from something I did manually. Once mesa is done compiling I will zap and try compile again to see if this package conflict still occurs.
